Panic in containerConfigFromError when running "buildx debug build"

### Contributing guidelines

- [x] I've read the [contributing guidelines](https://github.com/docker/buildx/blob/master/.github/CONTRIBUTING.md) and wholeheartedly agree

### I've found a bug and checked that ...

- [x] ... the documentation does not mention anything about my problem
- [x] ... there are no open or closed issues that are related to my problem

### Description

containerConfigFromError panics with buildx debug instead of handling error gracefully.

### Expected behaviour

Error should be handled gracefully as when running regular `docker buildx build`:

```console
cat <<EOF > spec.yml
# syntax=ghcr.io/project-dalec/dalec/frontend:0.20

name: go-md2man
version: 2.0.3
revision: 1
packager: Dalec Example
vendor: Dalec Example
license: MIT
description: A tool to convert markdown into man pages (roff).
website: https://github.com/cpuguy83/go-md2man


sources:
  src:
    generate:
      - gomod: {}
    git:
      url: https://github.com/cpuguy83/go-md2man.git
      commit: v2.0.3
  patch:
    context: {}
    includes:
      - patch

patches:
  src:
    - source: patch
      path: patch

dependencies:
  build:
    golang:

build:
  env:
    CGO_ENABLED: "0"
  steps:
    - command: |
        cd src
        go build -o go-md2man .

artifacts:
  binaries:
    src/go-md2man:

image:
  entrypoint: go-md2man
  cmd: --help

tests:
  - name: Check bin
    files:
      /usr/bin/go-md2man:
        permissions: 0o755
EOF
BUILDX_EXPERIMENTAL=1 docker buildx build --target azlinux3/container --progress plain -f ./spec.yml .
#0 building with "default" instance using docker driver

#1 [internal] load build definition from spec.yml
#1 transferring dockerfile: 868B done
#1 DONE 0.1s

#2 resolve image config for docker-image://ghcr.io/project-dalec/dalec/frontend:0.20
#2 DONE 0.2s

#3 docker-image://ghcr.io/project-dalec/dalec/frontend:0.20@sha256:60be1bad349ae49aeea971c0e46edee0c9f7926e4d62066e4e981c45289f6e10
#3 resolve ghcr.io/project-dalec/dalec/frontend:0.20@sha256:60be1bad349ae49aeea971c0e46edee0c9f7926e4d62066e4e981c45289f6e10 0.0s done
#3 CACHED

#4 resolve image config for docker-image://mcr.microsoft.com/azurelinux/base/core:3.0
#4 DONE 0.1s

#4 resolve image config for docker-image://mcr.microsoft.com/azurelinux/base/core:3.0
#4 DONE 0.2s

#5 [internal] load .dockerignore
#5 transferring context: 2B done
#5 DONE 0.0s

#4 resolve image config for docker-image://mcr.microsoft.com/azurelinux/base/core:3.0
#4 DONE 0.3s

#6 Generate RPM spec
#6 DONE 0.4s

#7 Prepare worker image
#7 resolve mcr.microsoft.com/azurelinux/base/core:3.0 0.2s done
#7 sha256:5546a28a785aeec63c8bd58a7dafb93b463e13f23d4fd7582a86d73722a6fb33 31.72MB / 31.72MB 8.0s done
#7 extracting sha256:5546a28a785aeec63c8bd58a7dafb93b463e13f23d4fd7582a86d73722a6fb33 0.9s done
#7 DONE 1.1s

#8 Prepare worker image
#8 CACHED

#9 Install RPMs
#9 CACHED

#10 Installing dependencies for go-md2man
#10 CACHED

#11 Fetch go module dependencies for source: src
#11 CACHED

#12 Create RPM buildroot
#12 CACHED

#13 Create RPM buildroot
#13 CACHED

#14 Prepare source: patch
#14 transferring context: 2B done
#14 DONE 0.0s

#15 Prepare source: patch
#15 CACHED

#16 Generate RPM spec
#16 CACHED

#17 Installing dependencies for go-md2man
#17 CACHED

#18 Build RPM
#18 CACHED

#19 Prepare source: src
#19 CACHED

#20 Generate RPM spec
#20 CACHED

#21 Add gomod sources
#21 ERROR: failed to calculate checksum of ref 4febh84oguow935h4r0q6gvf6::u79xwq39vocldssdtjcmysk2v: "/patch/patch": not found

#22 Prepare source: src
#22 CANCELED

#23 Generate RPM spec
#23 CANCELED

#24 Prepare worker image
#24 CANCELED
------
 > Add gomod sources:
------
spec.yml:15
--------------------
  14 |       src:
  15 | >>>     generate:
  16 | >>>       - gomod: {}
  17 | >>>     git:
  18 |           url: https://github.com/cpuguy83/go-md2man.git
--------------------
spec.yml:21
--------------------
  19 |           commit: v2.0.3
  20 |       patch:
  21 | >>>     context: {}
  22 |         includes:
  23 |           - patch
--------------------
spec.yml:27
--------------------
  25 |     patches:
  26 |       src:
  27 | >>>     - source: patch
  28 |           path: patch
  29 |
--------------------
ERROR: failed to build: failed to solve: failed to compute cache key: failed to calculate checksum of ref 4febh84oguow935h4r0q6gvf6::u79xwq39vocldssdtjcmysk2v: "/patch/patch": not found
```

### Actual behaviour

Panic occurred for me, which should likely be handled gracefully:
```console
ERROR: failed to compute cache key: failed to calculate checksum of ref 4febh84oguow935h4r0q6gvf6::c7p1abvstt1gr9kevmifumu84: "/keda-2.18-patch/specs/kedacore/patches/2.18.0/foo.patch": not found
[+] Building 0.0s (0/0)                                                                                                                                                                                                                          docker:default
Launching interactive container. Press Ctrl-a-c to switch to monitor console
Interactive container was restarted with process "kjo1s0bpzajd5i67wglgzwz44". Press Ctrl-a-c to switch to the new container
panic: runtime error: index out of range [0] with length 0

goroutine 218 [running]:
github.com/docker/buildx/build.containerConfigFromError(0x324409e0870, 0x32440bb6090)
	github.com/docker/buildx/build/result.go:225 +0x305
github.com/docker/buildx/build.(*ResultHandle).getContainerConfig(0x32440968cc0, 0x32440bb6090)
	github.com/docker/buildx/build/result.go:126 +0x216
github.com/docker/buildx/build.(*ResultHandle).NewContainer(0x32440968cc0, {0x55d01d4bb668, 0x324409553b0}, 0x10?)
	github.com/docker/buildx/build/result.go:76 +0x5c
github.com/docker/buildx/build.NewContainer.func1.1({0x55d01d4bb5f8, 0x55d01d764fa0}, 0x32440968cc0, 0x32440bb6090, 0x324405f5260)
	github.com/docker/buildx/build/invoke.go:82 +0x15a
github.com/docker/buildx/build.NewContainer.func1()
	github.com/docker/buildx/build/invoke.go:104 +0x30
created by github.com/docker/buildx/build.NewContainer in goroutine 248
	github.com/docker/buildx/build/invoke.go:77 +0x10c
```

This looks like `buildx debug` bug, not dalec bug, since it does not occur when running regular `buildx build`, which properly errors out.

### Buildx version

github.com/docker/buildx 0.33.0 f7897eba028583e0071642db3c011e860444f8cf

### Docker info

```text

```

### Builders list

```text
$ docker buildx ls
NAME/NODE     DRIVER/ENDPOINT   STATUS    BUILDKIT   PLATFORMS
default*      docker
 \_ default    \_ default       running   v0.29.0    linux/amd64 (+3), linux/arm64, linux/arm (+2), linux/ppc64le, (6 more)
```

### Configuration

```sh
cat <<EOF > spec.yml
# syntax=ghcr.io/project-dalec/dalec/frontend:0.20

name: go-md2man
version: 2.0.3
revision: 1
packager: Dalec Example
vendor: Dalec Example
license: MIT
description: A tool to convert markdown into man pages (roff).
website: https://github.com/cpuguy83/go-md2man


sources:
  src:
    generate:
      - gomod: {}
    git:
      url: https://github.com/cpuguy83/go-md2man.git
      commit: v2.0.3
  patch:
    context: {}
    includes:
      - patch

patches:
  src:
    - source: patch
      path: patch

dependencies:
  build:
    golang:

build:
  env:
    CGO_ENABLED: "0"
  steps:
    - command: |
        cd src
        go build -o go-md2man .

artifacts:
  binaries:
    src/go-md2man:

image:
  entrypoint: go-md2man
  cmd: --help

tests:
  - name: Check bin
    files:
      /usr/bin/go-md2man:
        permissions: 0o755
EOF
BUILDX_EXPERIMENTAL=1 docker buildx debug --on=error build --target azlinux3/container -f ./spec.yml .
```

### Build logs

```text

```

### Additional info

I'll open a PR with proposed fix soon.

--- a/build/result.go
+++ b/build/result.go
@@ -220,16 +220,24 @@ func containerConfigFromError(solveErr *errdefs.SolveError, cfg *InvokeConfig) (
 	if err != nil {
 		return nil, err
 	}
+	// When a solve fails before its inputs/mounts have been resolved (for
+	// example, when computing the cache key for an exec op fails because an
+	// input source cannot be found), MountIDs/InputIDs may be empty or shorter
+	// than the declared mount list. Guard against that so callers (such as the
+	// debug adapter) get a clean error instead of an index-out-of-range panic.
+	ids := solveErr.MountIDs
+	if cfg.Initial {
+		ids = solveErr.InputIDs
+	}
+	if len(ids) < len(exec.Mounts) {
+		return nil, errors.Errorf("cannot start debug container: failing step has no mount results available (got %d of %d)", len(ids), len(exec.Mounts))
+	}
 	var mounts []gateway.Mount
 	for i, mnt := range exec.Mounts {
-		rid := solveErr.MountIDs[i]
-		if cfg.Initial {
-			rid = solveErr.InputIDs[i]
-		}
 		mounts = append(mounts, gateway.Mount{
 			Selector:  mnt.Selector,
 			Dest:      mnt.Dest,
-			ResultID:  rid,
+			ResultID:  ids[i],
 			Readonly:  mnt.Readonly,
 			MountType: mnt.MountType,
 			CacheOpt:  mnt.CacheOpt,

--- a/tests/debug.go
+++ b/tests/debug.go
@@ -1,0 +1,131 @@
+package tests
+
+import (
+	"bytes"
+	"io"
+	"os/exec"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/containerd/continuity/fs/fstest"
+	"github.com/creack/pty"
+	"github.com/moby/buildkit/util/testutil/integration"
+	"github.com/stretchr/testify/require"
+)
+
+var debugTests = []func(t *testing.T, sb integration.Sandbox){
+	testDebugBuildMissingBindMountSource,
+}
+
+func testDebugBuildMissingBindMountSource(t *testing.T, sb integration.Sandbox) {
+	if !isExperimental() {
+		t.Skip("debug command is experimental")
+	}
+	if !isDockerWorker(sb) {
+		t.Skip("debug monitor test only needs a docker worker")
+	}
+
+	dockerfile := []byte(`
+FROM busybox:latest
+RUN --mount=type=bind,source=missing,target=/src true
+`)
+	dir := tmpdir(t,
+		fstest.CreateFile("Dockerfile", dockerfile, 0o600),
+	)
+
+	cmd := buildxCmd(sb, withArgs("debug", "--on=error", "build", "--progress=plain", "--output=type=cacheonly", dir))
+	f, err := pty.StartWithSize(cmd, &pty.Winsize{
+		Cols: 120,
+		Rows: 24,
+	})
+	require.NoError(t, err)
+	defer f.Close()
+
+	var output debugOutput
+	copyDone := make(chan struct{})
+	go func() {
+		_, _ = io.Copy(&output, f)
+		close(copyDone)
+	}()
+
+	waitCh := make(chan error, 1)
+	go func() {
+		waitCh <- cmd.Wait()
+	}()
+
+	exited, waitErr := waitForDebugPromptOrExit(t, &output, waitCh)
+	if !exited {
+		_, err = f.Write([]byte("exit\r"))
+		require.NoError(t, err)
+		waitErr = waitForDebugCommandExit(t, cmd, waitCh, &output)
+	}
+
+	_ = f.Close()
+	select {
+	case <-copyDone:
+	case <-time.After(time.Second):
+	}
+
+	out := output.String()
+	require.Error(t, waitErr, out)
+	require.Contains(t, out, "failed to calculate checksum")
+	require.Contains(t, out, "failed to create container")
+	require.NotContains(t, out, "panic:")
+	require.NotContains(t, out, "index out of range")
+}
+
+func waitForDebugPromptOrExit(t *testing.T, output *debugOutput, waitCh <-chan error) (bool, error) {
+	t.Helper()
+
+	timeout := time.NewTimer(time.Minute)
+	defer timeout.Stop()
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case err := <-waitCh:
+			return true, err
+		case <-ticker.C:
+			if strings.Contains(output.String(), "(buildx) ") {
+				return false, nil
+			}
+		case <-timeout.C:
+			require.FailNow(t, "timeout waiting for debug monitor", output.String())
+		}
+	}
+}
+
+func waitForDebugCommandExit(t *testing.T, cmd *exec.Cmd, waitCh <-chan error, output *debugOutput) error {
+	t.Helper()
+
+	select {
+	case err := <-waitCh:
+		return err
+	case <-time.After(30 * time.Second):
+		if cmd.Process != nil {
+			_ = cmd.Process.Kill()
+		}
+		require.FailNow(t, "timeout waiting for debug command to exit", output.String())
+		return nil
+	}
+}
+
+type debugOutput struct {
+	mu  sync.Mutex
+	buf bytes.Buffer
+}
+
+func (b *debugOutput) Write(p []byte) (int, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.buf.Write(p)
+}
+
+func (b *debugOutput) String() string {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.buf.String()
+}

--- a/tests/integration_test.go
+++ b/tests/integration_test.go
@@ -24,6 +24,7 @@ func TestIntegration(t *testing.T) {
 	var tests []func(t *testing.T, sb integration.Sandbox)
 	tests = append(tests, commonTests...)
 	tests = append(tests, buildTests...)
+	tests = append(tests, debugTests...)
 	tests = append(tests, policyBuildTests...)
 	tests = append(tests, policyEvalTests...)
 	tests = append(tests, policyBakeTests...)

